src/libNetwork/DataSender.cpp

```diff
@@ -0,0 +1,255 @@
+ */
+
+#include "DataSender.h"
+
+#include "libCrypto/Sha2.h"
+#include "libNetwork/P2PComm.h"
+#include "libUtils/DataConversion.h"
+#include "libUtils/Logger.h"
+
+using namespace std;
+
+void SendDataToLookupNodesDefault(const VectorOfLookupNode& lookups,
+                                  const vector<unsigned char>& message) {
+  if (LOOKUP_NODE_MODE) {
+    LOG_GENERAL(WARNING,
+                "DataSender::SendDataToLookupNodesDefault not "
+                "expected to be called from LookUp node.");
+  }
+  LOG_MARKER();
+
+  // TODO: provide interface in P2PComm instead of repopulating the lookup into
+  // vector of Peer
+  vector<Peer> allLookupNodes;
+
+  for (const auto& node : lookups) {
+    LOG_GENERAL(INFO, "Sending msg to lookup node " << node.second);
+
+    allLookupNodes.emplace_back(node.second);
+  }
+
+  P2PComm::GetInstance().SendBroadcastMessage(allLookupNodes, message);
+}
+
+void SendDataToShardNodesDefault(const vector<unsigned char>& message,
+                                 const DequeOfShard& shards,
+                                 const unsigned int& my_shards_lo,
+                                 const unsigned int& my_shards_hi) {
+  if (LOOKUP_NODE_MODE) {
+    LOG_GENERAL(WARNING,
+                "DataSender::SendDataToShardNodesDefault not expected to "
+                "be called from LookUp node.");
+    return;
+  }
+  // Too few target shards - avoid asking all DS clusters to send
+  LOG_MARKER();
+
+  auto p = shards.begin();
+  advance(p, my_shards_lo);
+
+  for (unsigned int i = my_shards_lo; i < my_shards_hi; i++) {
+    if (BROADCAST_GOSSIP_MODE) {
+      // Choose N other Shard nodes to be recipient of final block
+      std::vector<Peer> shardReceivers;
+      unsigned int numOfReceivers =
+          std::min(NUM_GOSSIP_RECEIVERS, (uint32_t)p->size());
+
+      for (unsigned int i = 0; i < numOfReceivers; i++) {
+        const auto& kv = p->at(i);
+        shardReceivers.emplace_back(std::get<SHARD_NODE_PEER>(kv));
+      }
+
+      P2PComm::GetInstance().SendRumorToForeignPeers(shardReceivers, message);
+    } else {
+      vector<Peer> shard_peers;
+
+      for (const auto& kv : *p) {
+        shard_peers.emplace_back(std::get<SHARD_NODE_PEER>(kv));
+        LOG_GENERAL(INFO, " PubKey: " << DataConversion::SerializableToHexStr(
+                                             std::get<SHARD_NODE_PUBKEY>(kv))
+                                      << " IP:Port "
+                                      << std::get<SHARD_NODE_PEER>(kv));
+      }
+
+      P2PComm::GetInstance().SendBroadcastMessage(shard_peers, message);
+    }
+
+    p++;
+  }
+}
+
+SendDataToLookupFunc SendDataToLookupFuncDefault =
+    [](const VectorOfLookupNode& lookups,
+       const std::vector<unsigned char>& message) mutable -> void {
+  SendDataToLookupNodesDefault(lookups, message);
+};
+
+SendDataToShardFunc SendDataToShardFuncDefault =
+    [](const std::vector<unsigned char>& message, const DequeOfShard& shards,
```

Comment:
It seems like `message` is unused for any of the actual instances.  Should we just remove `message` from the definition?

Reply:
It's used when P2P Sending Message